README.md
KEGS - Kent's Emulated GS
=========================

Kegs is an Apple //gs emulator for Mac OS X, Win32, Linux and Unix/X11

The latest version on the main KEGS home page is v0.91 which doesn't work on Intel Macs. In fact, only
small build changes are required, so I have forked the code to Github with my changes.

Original homepage with useful info on how to get it running is here: http://kegs.sourceforge.net/

Building
--------

* Install Apple's dev tools
* cd src
* make
* make to_pro # this is a utility for formatting disk images - see below

You will now have a .app for the emulator and a commandline tool to_pro in the src directory.

Disk image and compression utilities
------------------------------------

AppleCommander (Java): http://applecommander.sourceforge.net/
(NB: you need to be a little creative and run the jar at the commandline to work on current OSX)

NuLib: http://www.nulib.com/
Ciderpress (Windows): http://ciderpress.sourceforge.net/

See Also
--------

Sweet16 is a more capable emulator that seems to work on Lion out of the box: http://www.sheppyware.net/software-mac/sweet16/

License
-------

GPL (see COPYING.txt)

Original README.kegs (unedited atm, so some of this will be wrong)
==================================================================


KEGS: Kent's Emulated GS version 0.91
http://kegs.sourceforge.net/

What is this?
-------------

KEGS is an Apple IIgs emulator for Mac OS X, Linux, and Win32.
The Apple IIgs was the most powerful computer in the Apple II line.
It first was sold in 1986.  An Apple IIgs has the capability to run almost all
Apple II, Apple IIe, and Apple IIc programs.

KEGS supports all Apple IIgs graphics modes (which include all Apple //e
modes), plus plays all Apple IIgs sounds accurately.  It supports
serial port emulation through TCP/IP connections, or can use real
serial ports on Windows and Mac OS X.

The ROMs and GS/OS (the Apple IIgs operating system) are not included
with KEGS since they are not freely distributable.  KEGS is a little
user-hostile now, so if something doesn't work, let me know what went
wrong, and I'll try to help you out.  See my email address at the end of
this file.

I'd like to thank Chea Chee Keong who created KEGS32, the first Windows
port of KEGS.  That version, which has a better Windows-interface but which
is based on older core code, is at http://www.geocities.com/akilgard/kegs2.

KEGS features:
-------------

Fast 65816 emulation:
	About 80MHz on a Pentium 4 1.7GHz or a Mac G4 1GHz.
Emulates low-level 5.25" and 3.5" drive accesses (even nibble-copiers work!).
Emulates classic Apple II sound and 32-voice Ensoniq sound.
	All sound is played in 16-bit stereo at 48KHz (44100 on a Mac).
Emulates all Apple IIgs and Apple II graphics modes, including border effects.
	Can handle display changes at any time (superhires at the top, lores
	at the bottom).  Always does 60 full screen video updates per second.
	Supports 3200-color pictures.
Mouse and joystick support.
Emulated battery RAM remembers control panel settings.
Limited SCC (serial port) emulation to enable PR#1/2 IN#1/2 and Virtual
	Modem enables standard Apple terminal programs to telnet to any
	internet address (or receive connections from another telnet).

KEGS by default emulates a 8MB Apple IIgs, but you can change this from
the Configuration Panel.

KEGS is so accurate, even the built-in ROM selftests pass (you must be in
2.8MHz speed mode to pass the self-tests and you must set the Configuration
Panel entry "Enable Text Page 2 shadow" to "Disabled on ROM 01" for ROM 01).

Release info:
------------

Included files:
	CHANGES			- Description of changes since last release
	README.kegs.txt		- you're here
	README.compile.txt	- Describes how to build KEGS
	README.linux.rpm.txt	- Describes how to install KEGS's RPM for Linux
	README.win32.txt	- Win32 special directions
	README.mac.txt		- Mac OS X special directions
	README.a2.compatibility.txt - List of programs which need tweaking
	src/INTERNALS.overview	- description of how KEGS code works
	src/INTERNALS.xdriver	- Describes the xdriver.c routines for porting
	src/INTERNALS.iwm	- Describes the internal 3.5" and 5.25" disk
				   handling routines
	KEGSMAC			- the Mac OS X executable
	kegswin.exe		- the Windows executable
	config.kegs		- disk image configuration info
	to_pro			- Hard-to-use ProDOS volume creator
	partls			- Lists partitions on Apple-partitioned hard
				   drives or CD-ROMs
	src/			- All the source code, with a Makefile

You need to provide:

	1) Patience.
	2) a ROM file called "ROM", "ROM.01" or "ROM.03" in the KEGS directory
		(or in your home directory).  It can be either from a ROM 01
		(131072 bytes long) or from a ROM 03 machine (262144 bytes
		long.)
	3) A disk image to boot.  This can be either "raw" format or 2IMG.
		See discussion below.  GS/OS would be best.

Getting ROMs
------------

You need a copy of the memory from fe/0000 - ff/ffff from a ROM 01 GS
or fc/0000 - ff/ffff from a ROM 03 GS, and put that in a file called
"ROM".  I'll eventually write detailed instructions on how to do this.

Running KEGS:
------------

The distribution comes with the full source code for all platforms in
the src/ directory, the Windows executable as kegswin.exe, and the Mac OS X
executable as KEGSMAC.

See the README.compile.txt file for more info about compiling for Linux.

On all platforms except the Mac, you must start KEGS from a terminal
window.  KEGS will open a new window and use the window you started it from
as a "debug" window.  

On a MAC, you need to place the "config.kegs" file someplace where KEGS
can find it.  The simplest place is in your home directory, so copy it there
with the Finder (or using the Terminal).  You can also make the directory
Library/KEGS from your home directory, and then place config.kegs there
along with the ROM file.  You do not need a starting config.kegs file
on a Mac--KEGS will offer to make it for you if it cannot find one.

Start kegs by Double-clicking the KEGSMAC icon on a MAC, or by running
the executable (kegswin on Windows, and kegs on Linux).  KEGSMAC can
be run by the Terminal window on a Mac as well (which enables access to
more debug information) by typing: "./KEGSMAC.app/Contents/MacOS/KEGSMAC".

Assuming all goes well, KEGS will then boot up but probably not find any
disk images.  See below for how to tell KEGS what disk images to use.
Tip: Hitting "F8" locks the mouse in the window (and hides the host cursor)
until you hit "F8" again.

Configuration Panel:
-------------------

You enter the Configuration panel by pressing F4 at any time.  You tell
KEGS what disk images to use through the Configuration panel.  (If KEGS
couldn't find a ROM file, you will be forced into the Configuration
Panel mode until you select a valid ROM file).

To select a ROM file, select "ROM File Selection" and then select your
ROM file.  If you were not forced into the panel at startup, the KEGS
found one and you can skip this step.

Disk Images
-----------

The primary use of the Configuration Panel is to select disk images.  To
change disk images being used, select "Disk Configuration".  Each slot
and drive that can be loaded with an image is listed.  "s5d1" means slot
5, drive 1.  Slot 5 devices are 3.5" 800K disks, and slot 6 devices are
5.25" 140K disks.  Slot 7 devices are virtual hard drives, and can be
any size at all (although ProDOS-formatted images should be less than
32MB).

Just use the arrow keys to navigate to the device entry to change, and
then select it by pressing Return.  A scrollable file selection
interface is presented, letting you locate your image files.  To quickly
jump to a particular path, you can press Tab to toggle between entering
a path manually, and using the file selector. Press Return on ".."
entries to go up a directory level.  When you find the image you want,
just press Return.

If the image has partitions that KEGS supports, another selection
dialog will have you select which partition to mount.  You will probably
only have partitions on direct devices you mount (or on a Mac, of .dmg
images of CDs).  For instance, on a Mac, /dev/disk1 can sometimes be the
CDROM drive.

KEGS can handle "raw", .dsk, .po, 2IMG, 5.25" ".nib" images, most Mac
Diskcopy images and partitioned images.  The .dsk and .po formats you often
find on the web are really "raw" formats, and so they work fine.  KEGS uses
the host file permissions to encode the read/write status of the image.
KEGS can open any image file compressed with gzip (with the extension ".gz")
automatically as a read-only disk image.

An image is the representation of an Apple IIgs disk, but in a file on
your computer.  For 3.5" disks, for example, a raw image would be exactly
800K bytes long (819200 bytes).  KEGS directs the emulated GS accesses to
the image, and does the correct reads and writes of the Unix file instead.

To do "useful" things with KEGS, you need to get a bootable disk image.
You can go to http://www.info.apple.com/support/oldersoftwarelist.html and
get Apple IIgs System 6.  Unfortunately, Apple now only has .sea files which
are executable files for Macintosh only.  You need a macintosh to execute
those programs, which creates Disk Copy image files with no special extensions
(and with spaces in the names).  Once you get those files back to your
host machine, you can use them by selecting them from the Configuration Panel.

You can also get Apple II programs in ".dsk" format from a variety of
sites on the internet, and these should all work on KEGS as well.

KEGS also supports partitioned devices.  For instance, if you have a CD-ROM
on your computer, just pop an Apple II CD in, and KEGS can mount it, if
you have a Unix-based system (Linux, any Unix, and Mac OS X).

If you're on a Mac, be careful letting KEGS use your HFS partitions--
GSOS has many HFS bugs when it is writing.  Also avoid having KEGS access
an image which have mounted on your Mac at the same time (always unmount
it from your Mac before letting KEGS access it)!

If you do not have any disk mounted in s7d1, KEGS will jump into the monitor.
To boot slot 6 (or slot 5), use the Apple IIgs Control Panel by pressing
Ctrl-Command-ESC.

Support for 5.25" nibblized images is read-only for now (since the
format is kinda simplistic, it's tricky for KEGS to write to it since KEGS
has more information than fits in that format).  Just select your image,
like "disk.nib" in the kegs_conf file like any .dsk or .po image.

In addition to changing disks, you can also just "eject" and image by
moving the cursor to select that slot/drive and then press "E".  The
emulated IIgs will immediately detect changes to s5d1 and s5d2.

Care should be taken when changing images in slot 7--KEGS does not notify
GSOS that images have changed (or been ejected), and so it's best to make
changes when GSOS is not running.


Key summary:
-----------

F1:	Alias of Command
F2:	Alias of Option
F3:	Alias of ESC for OS/2 compatibility.
F4:	Configuration Panel
F6:	Toggle through the 4 speeds: Unlimited, 1MHz, 2.8MHz, 8.0MHz
Shift-F6: Enter KEGS debugger
F7:	Toggle fast_disk_emul on/off
F8:	Toggle pointer hiding on/off.
F9:	Invert the sense of the joystick.
Shift-F9: Swap x and y joystick/paddle axes.
F10:	Attempt to change the a2vid_palette (only useful on 256-color displays)
F11:	Full screen mode (only on Mac OS X).
F12:	Alias of Pause/Break which is treated as Reset

F2, Alt_R, Meta_r, Menu, Print, Mode_switch, Option:   Option key
F1, Alt_L, Meta_L, Cancel, Scroll_lock, Command:       Command key
Num_Lock:		Keypad "Clear".
F12, Pause, Break:	Reset

"Home": Alias for "=" on the keypad (since my Unix keyboard doesn't have an =).

Using KEGS:
----------

The host computer mouse is the Apple IIgs mouse and joystick by default.
By default, the host pointer is not constrained inside the window and
remains visible.  Press F8 to hide the cursor and constrain the mouse.  F8
again toggles out of constrain mode.  When the GSOS desktop is running,
KEGS hides the host cursor automatically and enables special tracking
which forces the emulated cursor to follow the host cursor.  If this doesn't
work right under some program, just press F8 for better compatibility.

The middle mouse button or Shift-F6 causes KEGS to stop emulation, and enter
the debugger.  You can continue with "g" then return in the debug window.
You can also disassemble memory, etc.  The section "Debugging KEGS"
above describes the debugger interface a little more.

KEGS has no pop-up menus or other interactive interfaces (other than
the debug window, and the occasional error dialogs on Mac OS X).  Input to
the debug window is only acted upon when the emulation is stopped
(Shift-F6, middle mouse button, or hitting a breakpoint).

Quitting KEGS:
-------------

Just close the main KEGS window, and KEGS will exit cleanly.  Or you
can select Quit from the menu.  Or enter ctrl-c in the debugger window.
Or press the middle-mouse button in the emulation window, and then type
"q" return in the debug window.

Command/Option keys:
-------------------

If you have a keyboard with the special Windows keys, you can
use them as the command/option keys.  For those without those keys,
there are several alternatives.

The following keys are Option (closed-apple) (not all keyboards have all
keys):  F2, Meta_R, Alt_R, Cancel, Print_screen, Mode_switch, Option,
or the Windows key just to the right of the spacebar.  The following keys are
Command (open-apple):  F1, Meta_L, Alt_L, Menu, Scroll_lock, Command,
the Windows key left of the spacebar, and the Windows key on the far right
that looks like a pull-down menu.  You can use F1 and F2 if you cannot make
anything else work (especially useful if your OS is intercepting some
Alt or Command key sequences).

If you can't get any of these to work on your machine, let me know.
Note that X Windows often has other things mapped to Meta- and Alt-
key sequences, so they often don't get passed through to KEGS.  So it's
best to use another key instead of Alt or Meta.

The joystick/paddle buttons are just the Command and Option keys.


Reset:
-----

The reset key is Pause/Break or F12.  You must hit it with Ctrl to get it to
take effect (just like a real Apple IIgs).  Ctrl-Command-Reset
forces a reboot.  Ctrl-Command-Option-Reset enters selftests.
Selftests will pass if you force speed to 2.8MHz using the middle
button or F6 (and also set Enable Text Page 2 shadow = Disabled for ROM 01).
Watch out for ctrl-shift-Break--it will likely kill an X Windows session.
Also note that the Unix olvwm X window manager interprets ctrl-F12 and will
not pass it on to KEGS--you'll need to use Break for reset in that case.

Full Screen mode (MAC OS X ONLY):
----------------

KEGS can run in full screen mode--which is especially useful when letting
small kids use KEGS (but it is not really a lock, so do not let a 2 year
old bang on the keyboard while running KEGS).

Full Screen mode is toggled with F11 (or Ctrl-F11, since Expose on a Mac
is intercepting F11).  If KEGS stops in the debugger for any reason,
full screen mode is toggled off automatically.


Joystick Emulation (Mouse, Keypad, or real native joystick):
------------------

The default joystick is the mouse position.  Upper left is 0,0.  Lower right
is 255,255.  Press Shift-F9 to swap the X and Y axes.  Press F9 to reverse
the sense of both paddles (so 0 becomes 255, etc).  Swapping and
reversing are convenient with paddle-based games like "Little Brick Out"
so that the mouse will be moving like the paddle on the screen.  "Little
Brick Out" is on the DOS 3.3 master disk.  The joystick does not work
properly if the pointer is constrained in the window.

You can also select from a "Keypad Joystick" or a real joystick from
the Configuration panel.  Press return on the "Joystick Configuration"
entry, and then select between Mouse Joystick, Keypad Joystick, or one
of two native joysticks.  The Keypad Joystick uses your keypad number
keys as a joystick, where keypad 7 means move to the upper left, and
keypad 3 means move to the lower right.  Pressing multiple keys together
averages the results, allowing finer control than just 8 directions.
Also, joystick scaling is selectable here for games which require
a greater range of motion to work correctly, along with trim adjustment
which moves the centering point.  Adjusting scaling usually means you
will need to adjust the trim as well.

The left mouse button is the mouse button for KEGS.  The right mouse
button (if you have it) or F6 toggles between four speed modes.  Mode 0
(the default) means run as fast as possible.  Mode 1 means run at 1MHz.
Mode 2 means run at 2.8MHz.  Mode 3 means run at 8.0MHz (about the speed
of a ZipGS accelerator).  Most Apple //e (or earlier) games need to be
run at 1MHz.  Many Apple IIgs demos must run at 2.8MHz or they will not
operate correctly.  Try running ornery programs at 2.8MHz.  3200 pictures
generally only display correctly at 2.8MHz or sometimes 8.0MHz.


Debugging KEGS:
--------------

KEGS by default now continues emulation even when it detects buggy programs
running.  (Now I know why Appleworks GS always seemed to crash!).

KEGS divides buggy programs into two severities: Code Yellow and Code Red.
The status is displayed in words in the text area under the emulation window.
If nothing's wrong, nothing is printed.

A Yellow bug is a mild bug where an Apple IIgs program merely read an
invalid location.  Although completely harmless, it indicates the potential
for some Apple IIgs program bug which may become more severe shortly.
For instance, closing the "About This Apple IIgs" window in the Finder
causes a code yellow alert, but it seems quite harmless.

A Code Red bug is a more serious problem.  The Apple IIgs program either
tried to write non-existent memory, entered an invalid system state, or
perhaps just tried to use an Apple IIgs feature which KEGS does not implement
yet.  Note that entering GSBUG tends to cause a Code Red alert always, so if
you intended to enter it, you can ignore it.  My recommendation is to
save work immediately (to new files) and restart KEGS if you get into the
Red mode.

KEGS also supports breakpoints and watchpoints.  In the debug window, you
set a breakpoint at an address by typing the address, followed by a 'B'
(it must be in caps).  To set a breakpoint on the interrupt jump point,
type:

e1/0010B

The format is "bank/address" then "B", where the B must be in caps and
the address must use lower-case hex.  For Apple IIe programs, just use a
bank of 0.

To list all breakpoints, just type 'B' with no number in front of it.
To delete a breakpoint, enter its address followed by 'D', so

e1/0010D

deletes the above breakpoint.  The addresses work like the IIgs monitor:
once you change banks, you can use shortcut addresses:

e1/0010B
14B

will add breakpoints at e1/0010 and e1/0014.

This is a "transparent" breakpoint--memory is not changed.  But any
read or write to that address will cause KEGS to halt.  So you can
set breakpoints on I/O addresses, or ROM, or whatever.  Setting a breakpoint
slows KEGS down somewhat, but only on accesses to the 256 byte "page"
the breakpoint is on. Breakpoints are not just instruction breakpoints,
they also cause KEGS to halt on any data access, too (usually called
watchpoints).

Frederic Devernay has written a nice help screen available in the
debugger when you type "h".

Useful locations for setting breakpoints:
0/3f0B		- Break handler
0/c000B		- Keyboard latch, programs read keys from this address



KEGS command-line option summary:
--------------------------------

There are others, but the Configuration panel provides a better way to
set them so they are no longer listed here.
-skip:	KEGS will "skip" that many screen redraws between refreshes.
	-skip 0 will do 60 frames per second, -skip 1 will do 30 fps,
	-skip 5 will do 10 fps.
-audio [0/1]: Forces audio [off/on].  By default, audio is on unless
	the X display is a remote machine or shared memory is off.
	This switch can override the default.  -audio 0 causes KEGS to
	not fork the background audio process, but Ensoniq emulation
	is still 100% accurate, just the sound is not sent to the
	workstation speaker.  Audio defaults off on Linux for now.
-arate {num}: Forces audio sample rate to {num}.  44100 and 48000 are
	usual, you can try 22050 to reduce KEGS's overhead.  On a reasonably

CHANGES

Changes in KEGS v0.86 since v0.85 (03/23/04)
- Add patch for Solaris sound by Jonathan Kalbfeld.
- Fix so that F4 enters config panel even while running Prosel-16
- Major mouse pointer changes, based on some ideas from Geoff Weiss.
	The GSOS mouse now exactly tracks the host pointer automatically.
- Fixed an accidental debug halt when Prosel-16 disables the keyboard/mouse.

Changes in KEGS v0.85 since v0.84 (01/09/04)
- Fix some minor 65816 bank-crossing bugs.
- Add -noignhalt to allow user to stop on code red halts.
- Fix Win32 capslock problem as reported by Edward Moore
- Fixed DreamVoir app on the sample image (it was corrupt)

Changes in KEGS v0.84 since v0.83 (11/21/03)
- Add new speed, 8.0MHz directly using right-clicking or F6.
- Sim speed and Video update interval added to Config panel.
- Various cycle timing bugs in engine_c.c fixed.
- Add Config Panel entry to mask serial output to 7-bit, to enable PR#2 to
	work better with an external telnet.
- In Config Panel file selection, typing a letter jumps to the first file
	beginning with that letter.
- Fixed various serial socket bugs.  Now you can disconnect a telnet session
	and start a new one, and a Linux hang is fixed.
- Default GS memory size increased to 8MB.
- Small fix to double-hires color table.
- X windows can now send displays to other-endian X servers.

Changes in KEGS v0.83 since v0.82 (11/19/03)
- Add Memory Size to config panel, with support for up to 14MB of memory
	(Geoff Weiss)
- Add $C04F EMUBYTE support which Bernie II the Rescue defined. (Geoff Weiss)
- Fix $CFFF code red's reported by David Wilson.
- Add smartport $C70A Format routine (David Wilson).

Changes in KEGS v0.82 since v0.81 (11/06/03)
- Fix superhires display glitch introduced in v0.81.
- Improved border handling--XMAS demo looks great.
- Fix some X build problems introduced in v0.81.

Changes in KEGS v0.81 since v0.80 (11/04/03)
- Code Red/Yellow warnings about emulation stability
- Windows file browsing fixes
- Built-in C600 ROM for Apple II 5.25" game compatibility
- Turns key repeat back on when exiting from X-windows version
- Windows F8 captures the cursor

Changes in KEGS v0.80 since v0.71 (10/31/03)
- Configuration Panel means no more hand-editing configuration files
- All emulator state is now saved in "config.kegs"
- 3200 color pictures!  Video system much improved for display accuracy.
- F8 Pointer grabbing works on Mac
- ZipGS emulation

Changes in KEGS v0.71 since v0.70 (11/20/02)
- Improved double-hires colors a lot.  -dhr140 is no longer the default
- Airheart relies on the PC going from 0xffff to 0x0000, so I undid the
	change from KEGS v0.54 which allowed PC to overflow to 0x10000.
	This slows KEGS down by about 5%.
- Fixed X shared memory bug in KEGS v0.70 with fix from Jonathan Stark.

Changes in KEGS v0.70 since v0.60 (11/18/02)
- New buttons: Middle button is enter-debugger, and right button changes speed
- New function key mapping (see README.kegs)
- Mac OS X port
- Win32 port
- Centralized much of what had been "xdriver.c" code into video.c, to move
	true platform-specific stuff into the various *driver.c codes.
	Kimage struct tracks video display buffers in a dev-independent way.
	From video.c, the calls to the platform code start with "x_" mostly.
	Code in video.c cleaned up somewhat.
	Borders are now always in native buffer format, while text/hires/
	and superhires are in 8-bit buffers and translated to native later.
- Mac and Windows sound are all done in one process--no child sound process.
- Revamped key press handling and mouse clicks--all is now handled in
	adb.c for a consistent user interface.  Now KEGS implements the
	same function keys on all platforms.  See README.kegs for fn key maps.
- I copied the debugger help from Frederic Devernay's KEGS-SDL port.
- Fixed an old IWM bug causing bad nibblization due to using uninit vars.
- Gilles Tschopp workaround to use corrupted 2IMG files (from KEGS-OSX).
- Gilles Tschopp provided code to zero //gs memory at startup (from KEGS-OSX)
- Simple code to try to use Mac Diskcopy format disks
- Ignore writes to 0xc0a8
- Search in $HOME and the launch directory (for mac) for kegs_conf/ROM
- Remove font65.sim file by integrating it into kegsfont.h.
- "-bw" option forces black and white hires mode.


Changes in KEGS v0.60 since v0.59 (10/03/00)
- The 16-bit colors were still wrong due to another coding error.  It would
	be much easier to get this right if I had a 16-bit color display...
	A user says it works now.

Changes in KEGS v0.59 since v0.58 (7/07/00)
- Added support for multiple paths to the default files and also multiple
	names for many default files.  This should make the .rpm distribution
	work better.
- Add another keycode to mean break according to mic@research.nj.nec.com.
- Add support for various ROMs to get plugged into slot 1-7.
- Fix code so that it should compile on 64-bit platforms.

Changes in KEGS v0.58 since v0.57 (2/08/00)
- Setting the execute bit on the disk image no longer means no-write-thru.
	Too many new users were getting confused by this.
- Fixed another bug with Apple //e bank switching created by v0.56
	Reported by phoenyx.
- Add command line option "-v" to turn on some verbose debugging flags.
- Fixed potential core-dump bug with non-8 bit visuals.
- Fixed double-lo-res color problem.
- The X driver should work with any visual depth display now and get the
	colors right.  Ian Schmidt reported his 16-bit card had bad colors.

Changes in KEGS v0.57 since v0.56 (12/27/99)
- Another try at making timezone stuff work across all Unix variants.
	Let me know if the time in the Apple //gs control panel doesn't
	match your real local time.
- Fix a bug created in v0.56 where the fast //e bank switch code had a typo.
	This prevented ZBasic from working correctly.

Changes in KEGS v0.56 since v0.55 (10/31/99)
- Faster Apple //e bank switch emulation.
- Simplified number of global variables for various softswitches.

src/clock.c
/************************************************************************/
/*			KEGS: Apple //gs Emulator			*/
/*			Copyright 2002 by Kent Dickey			*/
/*									*/
/*		This code is covered by the GNU GPL			*/
/*									*/
/*	The KEGS web page is kegs.sourceforge.net			*/
/*	You may contact the author at: kadickey@alumni.princeton.edu	*/
/************************************************************************/

const char rcsid_clock_c[] = "@(#)$KmKId: clock.c,v 1.29 2003-10-17 15:07:35-04 kentd Exp $";

#include "defc.h"
#include <time.h>
#ifdef _WIN32
# include <windows.h>
# include <mmsystem.h>
#else
# include <sys/time.h>
#endif

extern int Verbose;
extern int g_vbl_count;
extern int g_rom_version;
extern int g_config_kegs_update_needed;

#define CLK_IDLE		1
#define CLK_TIME		2
#define CLK_INTERNAL		3
#define CLK_BRAM1		4
#define CLK_BRAM2		5

int	g_clk_mode = CLK_IDLE;
int	g_clk_read = 0;
int	g_clk_reg1 = 0;

word32	c033_data = 0;
word32	c034_val = 0;

byte	g_bram[2][256];
byte	*g_bram_ptr = &(g_bram[0][0]);

word32	g_clk_cur_time = 0xa0000000;
int	g_clk_next_vbl_update = 0;

double
get_dtime()
{
#ifndef _WIN32
	struct timeval tp1;
	double	dsec;
	double	dusec;
#endif
	double	dtime;

	/* Routine used to return actual system time as a double */
	/* No routine cares about the absolute value, only deltas--maybe */
	/*  take advantage of that in future to increase usec accuracy */

#ifdef _WIN32
	dtime = timeGetTime() / 1000.0;
#else

# ifdef SOLARIS
	gettimeofday(&tp1, (void *)0);
# else
	gettimeofday(&tp1, (struct timezone *)0);
# endif

	dsec = (double)tp1.tv_sec;
	dusec = (double)tp1.tv_usec;

	dtime = dsec + (dusec / (1000.0 * 1000.0));
#endif

	return dtime;
}

int
micro_sleep(double dtime)
{
#ifndef _WIN32
	struct timeval Timer;
	int	ret;
#endif

	if(dtime <= 0.0) {
		return 0;
	}
	if(dtime >= 1.0) {
		halt_printf("micro_sleep called with %f!!\n", dtime);
		return -1;
	}

#if 0
	printf("usleep: %f\n", dtime);
#endif

#ifdef _WIN32
	Sleep(dtime * 1000);
#else
	Timer.tv_sec = 0;
	Timer.tv_usec = (dtime * 1000000.0);
	if( (ret = select(0, 0, 0, 0, &Timer)) < 0) {
		fprintf(stderr, "micro_sleep (select) ret: %d, errno: %d\n",
			ret, errno);
		return -1;
	}
#endif
	return 0;
}

void
clk_bram_zero()
{
	int	i, j;

	/* zero out all bram */
	for(i = 0; i < 2; i++) {
		for(j = 0; j < 256; j++) {
			g_bram[i][j] = 0;
		}
	}
	g_bram_ptr = &(g_bram[0][0]);
}

void
clk_bram_set(int bram_num, int offset, int val)
{
	g_bram[bram_num][offset] = val;
}

void
clk_setup_bram_version()
{
	if(g_rom_version < 3) {
		g_bram_ptr = (&g_bram[0][0]);	// ROM 01
	} else {
		g_bram_ptr = (&g_bram[1][0]);	// ROM 03
	}
}

void
clk_write_bram(FILE *fconf)
{
	int	i, j, k;

	for(i = 0; i < 2; i++) {
		fprintf(fconf, "\n");
		for(j = 0; j < 256; j += 16) {
			fprintf(fconf, "bram%d[%02x] =", 2*i + 1, j);
			for(k = 0; k < 16; k++) {
				fprintf(fconf, " %02x", g_bram[i][j+k]);
			}
			fprintf(fconf, "\n");
		}
	}
}

void
update_cur_time()
{
	struct tm *tm_ptr;
	time_t	cur_time;
	unsigned int secs, secs2;

	cur_time = time(0);

	/* Figure out the timezone (effectively) by diffing two times. */
	/* this is probably not right for a few hours around daylight savings*/
	/*  time transition */
	secs2 = mktime(gmtime(&cur_time));
	tm_ptr = localtime(&cur_time);
	secs = mktime(tm_ptr);

	secs = (unsigned int)cur_time - (secs2 - secs);

	if(tm_ptr->tm_isdst) {
		/* adjust for daylight savings time */
		secs += 3600;
	}

	/* add in secs to make date based on Apple Jan 1, 1904 instead of */
	/*   Unix's Jan 1, 1970 */
	/*  So add in 66 years and 17 leap year days (1904 is a leap year) */
	secs += ((66*365) + 17) * (24*3600);

	g_clk_cur_time = secs;

	clk_printf("Update g_clk_cur_time to %08x\n", g_clk_cur_time);
	g_clk_next_vbl_update = g_vbl_count + 5;
}

/* clock_update called by sim65816 every VBL */
void
clock_update()
{
	/* Nothing to do */
}

void
clock_update_if_needed()
{
	int	diff;

	diff = g_clk_next_vbl_update - g_vbl_count;
	if(diff < 0 || diff > 60) {
		/* Been a while, re-read the clock */
		update_cur_time();
	}
}

word32
clock_read_c033()
{
	return c033_data;
}

word32
clock_read_c034()
{
	return c034_val;
}

void
clock_write_c033(word32 val)
{
	c033_data = val;
}

void
clock_write_c034(word32 val)
{
	c034_val = val & 0x7f;
	if((val & 0x80) != 0) {
		if((val & 0x20) == 0) {
			printf("c034 write not last = 1\n");
			/* set_halt(1); */
		}
		do_clock_data();
	}
}


void
do_clock_data()
{
	word32	mask;
	int	read;
	int	op;

	clk_printf("In do_clock_data, g_clk_mode: %02x\n", g_clk_mode);

	read = c034_val & 0x40;
	switch(g_clk_mode) {
	case CLK_IDLE:
		g_clk_read = (c033_data >> 7) & 1;
		g_clk_reg1 = (c033_data >> 2) & 3;
		op = (c033_data >> 4) & 7;
		if(!read) {
			/* write */
			switch(op) {
			case 0x0:	/* Read/write seconds register */
				g_clk_mode = CLK_TIME;
				clock_update_if_needed();
				break;
			case 0x3:	/* internal registers */
				g_clk_mode = CLK_INTERNAL;
				if(g_clk_reg1 & 0x2) {
					/* extend BRAM read */
					g_clk_mode = CLK_BRAM2;
					g_clk_reg1 = (c033_data & 7) << 5;
				}
				break;
			case 0x2:	/* read/write ram 0x10-0x13 */
				g_clk_mode = CLK_BRAM1;
				g_clk_reg1 += 0x10;
				break;
			case 0x4:	/* read/write ram 0x00-0x0f */
			case 0x5: case 0x6: case 0x7:
				g_clk_mode = CLK_BRAM1;
				g_clk_reg1 = (c033_data >> 2) & 0xf;
				break;
			default:
				halt_printf("Bad c033_data in CLK_IDLE: %02x\n",
					c033_data);
			}
		} else {
			printf("clk read from IDLE mode!\n");
			/* set_halt(1); */
			g_clk_mode = CLK_IDLE;
		}
		break;
	case CLK_BRAM2:
		if(!read) {
			/* get more bits of bram addr */
			if((c033_data & 0x83) == 0x00) {
				/* more address bits */
				g_clk_reg1 |= ((c033_data >> 2) & 0x1f);
				g_clk_mode = CLK_BRAM1;
			} else {
				halt_printf("CLK_BRAM2: c033_data: %02x!\n",
						c033_data);
				g_clk_mode = CLK_IDLE;
			}
		} else {
			halt_printf("CLK_BRAM2: clock read!\n");
			g_clk_mode = CLK_IDLE;
		}
		break;
	case CLK_BRAM1:
		/* access battery ram addr g_clk_reg1 */
		if(read) {
			if(g_clk_read) {
				/* Yup, read */
				c033_data = g_bram_ptr[g_clk_reg1];
				clk_printf("Reading BRAM loc %02x: %02x\n",
					g_clk_reg1, c033_data);
			} else {
				halt_printf("CLK_BRAM1: said wr, now read\n");
			}
		} else {
			if(g_clk_read) {
				halt_printf("CLK_BRAM1: said rd, now write\n");
			} else {
				/* Yup, write */
				clk_printf("Writing BRAM loc %02x with %02x\n",
					g_clk_reg1, c033_data);
				g_bram_ptr[g_clk_reg1] = c033_data;
				g_config_kegs_update_needed = 1;
			}
		}
		g_clk_mode = CLK_IDLE;
		break;
	case CLK_TIME:
		if(read) {
			if(g_clk_read == 0) {
				halt_printf("Reading time, but in set mode!\n");
			}
			c033_data = (g_clk_cur_time >> (g_clk_reg1 * 8)) & 0xff;
			clk_printf("Returning time byte %d: %02x\n",
				g_clk_reg1, c033_data);
		} else {
			/* Write */
			if(g_clk_read) {
				halt_printf("Write time, but in read mode!\n");
			}
			clk_printf("Writing TIME loc %d with %02x\n",
				g_clk_reg1, c033_data);
			mask = 0xff << (8 * g_clk_reg1);

			g_clk_cur_time = (g_clk_cur_time & (~mask)) |
				((c033_data & 0xff) << (8 *g_clk_reg1));
		}
		g_clk_mode = CLK_IDLE;
		break;
	case CLK_INTERNAL:
		if(read) {
			printf("Attempting to read internal reg %02x!\n",
				g_clk_reg1);
		} else {
			switch(g_clk_reg1) {
			case 0x0:	/* test register */
				if(c033_data & 0xc0) {
					printf("Writing test reg: %02x!\n",
						c033_data);
					/* set_halt(1); */
				}
				break;
			case 0x1:	/* write protect reg */
				clk_printf("Writing clk wr_protect: %02x\n",
					c033_data);
				if(c033_data & 0x80) {
					printf("Stop, wr clk wr_prot: %02x\n",
						c033_data);
					/* set_halt(1); */
				}
				break;
			default:
				halt_printf("Writing int reg: %02x with %02x\n",
					g_clk_reg1, c033_data);
			}
		}
		g_clk_mode = CLK_IDLE;
		break;
	default:
		halt_printf("clk mode: %d unknown!\n", g_clk_mode);
		g_clk_mode = CLK_IDLE;
		break;
	}
}

